First-class support for PostgreSQL collations

roji opened the pull request:
Spent some more time on this, it sort of morphed out of control:

* Full support for global database collation and per-column, including reverse engineering.
* Migration management of collation database objects, allowing creating and dropping collations like other objects.
* EF.Functions.ApplyCollation allow explicit collations in queries.
* Added UseCase{Sensitive,Insensitive}Collation to the database and property levels. This defines which collation is used for explicitly case-{in,}sensitive operations; translating string.Equals(x, StringComparison) works with Ordinal and OrdinalIgnoresCase.

Closes #406
Replaces #1326

roji commented:
@dotnet/efteam here's a pretty complete implementation of a full collations story for PostgreSQL (e.g. https://github.com/dotnet/efcore/issues/1222, https://github.com/dotnet/efcore/issues/19866...). Here's a super-quick summary:

* There's metadata which manages creating columns with collations (e.g. `ALTER TABLE ""People"" ADD ""Name"" text NULL COLLATE "de_DE"`).
* There's an EF.Functions.ApplyCollation which translates into a CollationExpression in the tree, that gets generated to SQL such as `WHERE s.""CS"" COLLATE ""ci_coll"" = 'sometext'`.
* I went the extra mile and allowed users to also configure collations for "explicitly-defined case-sensitive (or insensitive) operations". If a user does this by calling UseCaseInsensitiveCollation on a property (or on the entire model), they can use `string.Equals(s, StringComparison.OrdinalIgnoresCase)`. This requires looking up property annotations in the model from with query translation, which I haven't seen us do yet (can you think of any problems with this?).
* There's also metadata for specifying collations on indexes so all this can be fast.

[Here are some tests for your viewing pleasure](https://github.com/npgsql/efcore.pg/blob/91c3e49d0411d266f065f46f32b471c3e53aae66/test/EFCore.PG.FunctionalTests/Query/CollationQueryTest.cs).

/cc @Suchiman

Suchiman commented:
This looks great!

I'm not entirely sold on `UseCollation` setting the column storage collation and `UseCaseInsensitiveCollation` specifying the comparer collation but naming this is really hard.
The best i could think of would be either `.UseCaseInsensitiveComparerCollation()` or `.CompareCaseInsensitiveCollation()`

smitpatel commented on test/EFCore.PG.FunctionalTests/Query/CollationQueryTest.cs:
```diff
@@ -0,0 +1,201 @@
+            using var ctx = CreateContext();
+
+            var id = ctx.SomeEntities
+                .Where(b => EF.Functions.ApplyCollation(b.CS, "ci_coll") == "sometext")
```
`EF.Functions.Collate`?

roji commented:
@lauxjpn you may be interested in this. I wonder how much of it is relevant to MySQL.

roji commented:
@Suchiman

> I'm not entirely sold on UseCollation setting the column storage collation and UseCaseInsensitiveCollation specifying the comparer collation but naming this is really hard.
The best i could think of would be either .UseCaseInsensitiveComparerCollation() or .CompareCaseInsensitiveCollation()

Yeah, I had a hard time with the naming and am still not very happy - will take another look at it soon. The naming for UseCaseInsensitiveCollation definitely doesn't seem ideal, but I'm not sure adding the word Comparer in there makes things a lot better...

Note that unfortunately PostgreSQL is somewhat limited in what can be defined as a database-wide collation (at database-creation time). So in order for this to be useful, I'll probably have to expose both a UseDatabaseCollation (which affects CREATE DATABASE like the PR's current UseCollation), and a UseDefaultColumnCollation, which is propagated by EF Core as a default when creating all columns.

lauxjpn commented:
> @lauxjpn you may be interested in this. I wonder how much of it is relevant to MySQL.

Definitely some intersting stuff in there, like the `DbFunctions` extension method or parts of the `VisitEqualsWithStringComparison()`.

In MySQL, database level collations will only set the default used for tables and columns.

So if `Collation A` is defined in `Migration A`, then `Table B` is introduced in `Migration B` and finally the database collation is changed to `Collation C` in `Migration C`, then `Table B`'s collation would still be `Collation A`.

If somebody would instead use the `DbContext.Database.EnsureCreate()` method however, he would end up with `Table B`'s collation being `Collation C`.

We have plans how to handle this issue (with some MySQL specific SQL statements), but it could be relevant to PostgreSQL as well.

smitpatel commented on src/EFCore.PG/Extensions/NpgsqlDbFunctionsExtensions.cs:
```diff
@@ -46,5 +46,24 @@ public static class NpgsqlDbFunctionsExtensions
+        public static T ApplyCollation<T>(
+            [NotNull] this DbFunctions _,
+            [NotNull] T operand,
+            [NotNull] string collation)
```
`[NotParameterized]`?

smitpatel commented on src/EFCore.PG/Query/Expressions/Internal/CollateExpression.cs:
```diff
@@ -0,0 +1,70 @@
+        public virtual string Collation { get; }
+
+        public CollateExpression([NotNull] SqlExpression operand, [NotNull] string collation)
+            : base(operand.Type, operand.TypeMapping)
```
Is this TypeMapping valid? Does the typeMapping print out literals in correction collation?

roji replied:
In this PR, the type mapping doesn't know anything about the collation, since it's not really part of the type name, but of the column spec (see https://github.com/dotnet/efcore/issues/11896#issuecomment-604849819). So with this PR, the provider only ever emits COLLATE:

1. In column specs DDL (e.g. when creating a new column)
2. When EF.Functions.ApplyCollation (soon to be Collate) is applied explicitly
3. When string.Equals(s, StringComparison.OrdinalIgnoresCase) is used (or Ordinal).

Therefore the type mapping doesn't seem to be concerned. Tell me if you see an issue with this approach.

smitpatel commented on src/EFCore.PG/Query/Internal/NpgsqlSqlTranslatingExpressionVisitor.cs:
```diff
@@ -383,5 +475,24 @@ bool TranslationFailed(Expression original, Expression translation, out SqlExpre
 
         #endregion Copied from RelationalSqlTranslatingExpressionVisitor
+
+        public class ExpressionTypeExtractingExpressionVisitor<T> : ExpressionVisitor
```
This type of visitors are bug prone.

roji replied:
Can you provide more detail on why? The basic need here is for the configured case-sensitive/insensitive collation to bubble up the expression tree (a bit like how type mappings bubble up), so that we can specify it explicitly at the comparison level.

smitpatel replied:
TypeMapping used to have similar bubbling up code in past (albeit via a recursive method). But what does bubbling should do?
In expression `a + b` if `a` has collation and `b` does not then what is the collation of resulting expression?
Places where things are complicated
Suppose there is some function `Concat(a,b,c)`
what is resulting collation? 
- If only a has collation or only b has collation.
- If a,b,c all has collation but same value
- If a,b,c all has collation but different value
Further for translation of expression pattern like `"abc".IndexOf("b").ToString()`
IndexOf returns an int which is converted to string. It should not have barring with collation of "abc" or "b" but visitor would bubble it up still.

roji replied:
The rule currently implemented, is that we expect exactly one collation to bubble out. If an expression (recursively) contains more than one column, and there are non-identical collations defined on those columns, we error ([see here](https://github.com/npgsql/efcore.pg/pull/1329/files#diff-c61aff25e85385e445833ca8538ef4acR234)). If there are no collations, we error as well (as today). This seems like a reasonable thing to do - do you disagree?

> Further for translation of expression pattern like "abc".IndexOf("b").ToString()
> IndexOf returns an int which is converted to string. It should not have barring with collation of "abc" or "b" but visitor would bubble it up still.

That's a good point. We could narrow things a bit, and say that we only bubble up the collation across operators/functions which also return a string (hence Concat would bubble up, but IndexOf wouldn't). I'm sure it's possible to find an example where this would be weird, but is there a better alternative? If we think this is too problematic, we could cut out translation of string.Equals with StringComparison, and require the user to explicitly define collations via EF.Functions.Collate - but there seems to be value in translating string.Equals and I'm not sure the actual problem is serious. What do you think?

PS this discussion seems very relevant also for SQL Server and MySQL support.

roji replied:
Of course, one more possibility is only to look for a collation on the immediate parameters of string.Equals(). This would limit the usefulness a bit, but sounds like it would cover what most people actually need.

@smitpatel and others, thoughts?

smitpatel replied:
It is an approximate solution. We could put guards in to reduce false cases like only expecting one collation to bubble out. We could also improve correctness by blocking (or bubbling up) of certain operations. Or we could just not do it. At the end it would depend on how much cases it solves and what is the accuracy.
For typemapping it just did not work well for us. It had too many negative cases. At the same time, type mapping is a lot more widespread compared to this where you need only when translating string.Equals with Comparison. So it is not entirely bad. Also depends how much people are going to do complex operation before doing this kind of comparison. Most comparisons would be column to column/constant/parameter match only.
EF.Functions.Collate would always be an escape hatch for any implementation we add here (even if we don't try to find any expression this way and just use collation from CollateExpression generated by the function translation).
On two extremes
- We could keep this as is. May cause incorrect matches. Use escape hatch. Not sure how easy would it to identify incorrect match.
- We could remove this visitor, match probably directly CollateExpression or direct ColumnExpression. This should cover most basic cases. Relative use of basic case vs complex case is unknown to me.

An in between solution would be to let it go through very specific operations we know are safe. Like concat `string.Equals(a + b, c, ...)` Though we would need to define what would be behavior when only one of a/b has collation or both have different values. (Could also consider ScalarSubqueryExpression in this. Essentially form of `c.Orders.Select(...).FirstOrDefault()`). May be few more?
In any case where it did not work, translation would fail and user can use Collate API.

Note: When looking at collation, just like typeMapping, ignore the facet for constant/parameter. So a + constant should have same collation as a.

smitpatel commented:
Agree with @Suchiman - The naming of case sensitive/insensitive is quite confusing. I had to go back & forth to understand the difference. Comparer or Comparison in name could specify more details perhaps.

roji commented on src/EFCore.PG/Design/Internal/NpgsqlAnnotationCodeGenerator.cs:
```diff
@@ -119,6 +119,9 @@ public override MethodCallCodeFragment GenerateFluentApi(IModel model, IAnnotati
             }
 
+            if (annotation.Name == NpgsqlAnnotationNames.Collation)
+                return new MethodCallCodeFragment(nameof(NpgsqlModelBuilderExtensions.UseCollation));
```
Good catch, thanks!

roji commented:
> Agree with @Suchiman - The naming of case sensitive/insensitive is quite confusing. I had to go back & forth to understand the difference. Comparer or Comparison in name could specify more details perhaps.

OK, will change the naming. I guess I'm leaning towards UseCaseInsensitiveComparisonCollation (although it still seems a bit silly since collation is only about comparison in any case).

roji commented:
@lauxjpn 

> In MySQL, database level collations will only set the default used for tables and columns.

In PostgreSQL as well, unless I'm mistaken the database-level collation is basically only the implicit default for columns being created. Does MySQL have a notion of *table* collation? What does that mean? :)

> So if Collation A is defined in Migration A, then Table B is introduced in Migration B and finally the database collation is changed to Collation C in Migration C, then Table B's collation would still be Collation A.

That makes sense to me. Note that in PostgreSQL, the database collation (which again is the default for all columns) cannot be changed after the database is created, and is also somewhat limited in what collations can be specified. As a result, I'll probably introduce a separate database-level annotation which acts as an *EF Core* column default, instead of setting the PostgreSQL default (see [this comment](https://github.com/npgsql/efcore.pg/pull/1329#issuecomment-608148363)). In other words, instead of specifying the collation when creating the database, the provider would specify it on each and every column that gets created. Besides solving the PostgreSQL-specific issues with database-wide collation, this would also not have the issue you mentioned - if a user ever changes this annotation, the next migration would automatically alter the collations of all columns (I need to test this works). That would work very similarly to how the value generation strategy can be defined at the database level, and is propagated everywhere.

roji commented:
Pushed some more work on this, addressing review comments.

We also now have two ways to specify collations at the model level:

1. UseDatabaseCollation, which affects the `CREATE DATABASE` statement. The database-wide collation cannot be altered once the database is created, and only supports libc collations and not ICU, making this rather limited.
2. UseDefaultColumnCollation, which works as a model-wide EF Core default similar to HasValueGenerationStrategy: it makes all columns get created with an explicit collation.

Note: I haven't renamed UseCase{Sensitive,Insensitive}Collation, let's wait for the output of https://github.com/npgsql/efcore.pg/pull/1329#discussion_r402840991 to see if we're keeping this feature.

roji commented:
Based on the work and the feedback here, I submitted support upstream in EF Core (#20602, #20603). I'll close this PR, but once those changes make it back via upstream I'll submit a separate PR for only the remaining PostgreSQL-specific aspects.

mhosman commented:
Great! This is working with EFCore 3.1? Do you have an example of how to use UseCaseInsensitiveComparisonCollation at database creation? Another question, there's something like this to ignore accents? Thanks!!!

roji commented:
@mhosman another version of this has been merged for 5.0 - not 3.1. Specifically, UseCaseInsensitiveComparisonCollation isn't implemented - but if you need to specify a case-insensitive collation at the column level that will be possible (I still have to merge some special PG support for doing this at the model level too).

To get an idea of the general EF support that's coming for collations, [look at these docs](https://github.com/dotnet/EntityFramework.Docs/pull/2328/files). But as I said, there's another small layer of PG-specific functionality which I have yet to document.
